helix-term/src/commands.rs

```diff
@@ -4014,16 +4013,50 @@ pub mod insert {
         doc.apply(&transaction, view.id);
     }
 
-    // TODO: handle indent-aware delete
     pub fn delete_char_backward(cx: &mut Context) {
         let count = cx.count();
         let (view, doc) = current!(cx.editor);
         let text = doc.text().slice(..);
         let transaction =
             Transaction::change_by_selection(doc.text(), doc.selection(view.id), |range| {
                 let pos = range.cursor(text);
+                let line_start_pos = text.line_to_char(range.cursor_line(text));
+
+                // considier to delete by indent level if all characters before `pos` are indent units.
+                let indent_unit = doc.indent_unit();
+                let line_to_start = text.slice(line_start_pos..pos).as_str().unwrap();
+                // using unwrap here is ok because indent_unit always contains `spaces` or `tab`.
+                let indent_unit_base = indent_unit.chars().nth(0).unwrap();
+
+                // it's ok to dedent when there are only `indent_unit_base` character(e.g: ' ') in `line_to_start`.
+                let mut ok_to_dedent = true;
+                for c in line_to_start.chars() {
+                    if c != indent_unit_base {
+                        ok_to_dedent = false;
+                        break;
+                    }
+                }
+
+                let delete_count = if ok_to_dedent {
+                    // current position just at start of line.
+                    if line_to_start.is_empty() {
+                        count
+                    } else {
+                        let steps_to_aligned = line_to_start.len() % indent_unit.len();
+                        // current position is just aligned with indent_unit length, when hit delete, should
+                        // delete a whole `indent_unit`.
+                        if steps_to_aligned == 0 {
+                            indent_unit.len()
+                        } else {
+                            steps_to_aligned
+                        }
+                    }
+                } else {
+                    count
+                };
```

Comment:
Let's instead use the codemirror algorithm which can handle mixed indentation:

https://github.com/codemirror/commands/blob/ad908aaf12c84798a6e6dab5d0d770437abe5a40/src/commands.ts#L383-L388

https://github.com/codemirror/text/blob/main/src/column.ts#L3-L17

Pseudocode in Rust:

```rust
use helix_core::unicode::width::UnicodeWidthChar;

let fragment = Cow::from(text.slice(line_start_pos..pos));
if fragment.chars().all(|ch| ch.is_whitespace()) {
   if get_char(pos.saturating_sub(1)) == Some('\t') {
      // fast path, delete one char
   } else {
     let unit_len = indent_unit.chars().count();
     let unit_size = if indent_unit.starts_with('\t') { tab_size * unit_len } else { unit_len };
     let width = fragment.chars().map(|ch| if ch == '\t' { tab_size } else { ch.width() }).sum();
     let drop = width % unit_size;  // round down to nearest unit
     if drop == 0 { drop = unit_size };  // if it's already at a unit, consume a whole unit
     let mut chars = text.chars_at(pos).rev();
     let mut start = pos;
     for i in 0..drop { // delete up to `drop` spaces
        match chars.next() {
          Some(' ') => start -= 1,
          None => break,
        }
     }
     (start, pos, None) // delete!
   }
} else {
   // delete char
}
```

Reply:
Done with little adjusted code :)